File: better-ddd/src/test/java/com/example/FieldAccessAspect.java
package com.example;

import org.aspectj.lang.JoinPoint;
import org.aspectj.lang.ProceedingJoinPoint;
import org.aspectj.lang.annotation.Around;
import org.aspectj.lang.annotation.Aspect;
import org.aspectj.lang.annotation.Before;
import org.aspectj.lang.annotation.Pointcut;
import org.aspectj.lang.reflect.FieldSignature;

import java.lang.reflect.Field;

@Aspect
public class FieldAccessAspect {

    @Pointcut("get(* *) && @within(org.morecup.jimmerddd.betterddd.annotation.AggregateRoot) && !within(com.example.FieldAccessAspect)")
    public void anyFieldGet() {}

    @Pointcut("set(* *) && @within(org.morecup.jimmerddd.betterddd.annotation.AggregateRoot) && !within(com.example.FieldAccessAspect)")
    public void anyFieldSet() {}

    @Around("anyFieldGet()")
    public Object aroundFieldGet(ProceedingJoinPoint pjp) throws Throwable {
        FieldSignature fieldSignature = (FieldSignature) pjp.getSignature();
        Field field = fieldSignature.getField();

        Object originalValue = pjp.proceed();
        System.out.println("Intercepted read field " + ((FieldSignature) pjp.getSignature()).getFieldType() + ", original value: " + originalValue);

        if (originalValue instanceof String) {
            return originalValue + "-intercepted";
        }
        return originalValue;
    }

    @Around("anyFieldSet()")
    public void aroundFieldSet(ProceedingJoinPoint pjp) throws Throwable {
        Object[] args = pjp.getArgs();
        Object originalValue = args[0];
        System.out.println("Intercepted write field " + pjp.getSignature() + ", original value: " + originalValue);

        Object newValue = originalValue;
        if (originalValue instanceof String) {
            newValue = "Intercepted-" + originalValue;
        }

        pjp.proceed(new Object[]{newValue});
    }
}

File: better-ddd/src/main/kotlin/org/morecup/jimmerddd/betterddd/aspect/FieldAccessAspect.kt
package org.morecup.jimmerddd.betterddd.aspect

import org.aspectj.lang.ProceedingJoinPoint
import org.aspectj.lang.annotation.Around
import org.aspectj.lang.annotation.Aspect
import org.aspectj.lang.annotation.Pointcut
import org.aspectj.lang.reflect.FieldSignature
import java.lang.reflect.Field

@Aspect
class FieldAccessAspect {

    @Pointcut("get(* *) && @within(org.morecup.jimmerddd.betterddd.annotation.AggregateRoot) && !within(org.morecup.jimmerddd.betterddd.aspect.FieldAccessAspect)")
    fun anyFieldGet() {}

    @Pointcut("set(* *) && @within(org.morecup.jimmerddd.betterddd.annotation.AggregateRoot) && !within(org.morecup.jimmerddd.betterddd.aspect.FieldAccessAspect)")
    fun anyFieldSet() {}

    @Around("anyFieldGet()")
    fun aroundFieldGet(pjp: ProceedingJoinPoint): Any {
        val fieldSignature = pjp.signature as FieldSignature
        val field = fieldSignature.field

        val originalValue = pjp.proceed()
        println("Intercepted read field ${fieldSignature.fieldType}, original value: $originalValue")

        return if (originalValue is String) {
            "$originalValue-intercepted"
        } else {
            originalValue
        }
    }

    @Around("anyFieldSet()")
    fun aroundFieldSet(pjp: ProceedingJoinPoint) {
        val args = pjp.args
        val originalValue = args[0]
        println("Intercepted write field ${pjp.signature}, original value: $originalValue")

        val newValue = if (originalValue is String) {
            "Intercepted-$originalValue"
        } else {
            originalValue
        }

        pjp.proceed(arrayOf(newValue))
    }
}